### PR TITLE
[RW-6725][risk=low] Real access renewal status in profile response

### PR DESCRIPTION
**Pull request by dmohs**
I tried to make this as concise as possible and would appreciate pointers. Unit tests are a big lift, so added RW-6750 to cover them.

I judge risk to be low because getting the logic wrong could cause someone to stay registered longer than they should be. Avoiding this should be covered by existing tests, but I haven't checked.

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 1]
```diff
@@ -211,30 +213,92 @@ private void updateUserAccessTiers(DbUser dbUser, Agent agent) {
...
 
-  public boolean isCompleteAndNotExpired(Timestamp completionTime) {
-    Timestamp expirationTime =
-        new Timestamp(
-            clock.millis()
-                - TimeUnit.MILLISECONDS.convert(
-                    configProvider.get().accessRenewal.expiryDays, TimeUnit.DAYS));
-    if (configProvider.get().access.enableAccessRenewal) {
-      return completionTime != null && expirationTime.before(completionTime);
+  private class ModuleTimes {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:216]: I like this.  Probably makes sense to break it out into its own file.   Maybe under the access package (where I hope all this code will live eventually)

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 2]
```diff
@@ -211,30 +213,92 @@ private void updateUserAccessTiers(DbUser dbUser, Agent agent) {
...
+      long expiryDaysInMs = TimeUnit.MILLISECONDS.convert(expiryDays, TimeUnit.DAYS);
+      return Optional.of(new Timestamp(completion.get().getTime() + expiryDaysInMs));
+    }
+
+    public boolean hasExpired() {
+      final Timestamp now = new Timestamp(clock.millis());
+      return getExpiration().map(x -> x.before(now)).orElse(false);
+    }
+
+    public boolean areCompliant() {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:256]: this is not used

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 3]
```diff
@@ -211,30 +213,92 @@ private void updateUserAccessTiers(DbUser dbUser, Agent agent) {
...
+
+    public boolean isBypassed() {
+      return bypass.isPresent();
+    }
+
+    public boolean isComplete() {
+      return completion.isPresent();
+    }
+
+    public Optional<Timestamp> getExpiration() {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:233]: we can shrink this a bit by reordering and using Optional.map().  Here is my thought:

if (isBypassed() || !config.enableRenewal) return empty;

if (expiryDays == null) throw ServerError/500 or something

return completion.map(c -> new Timestamp(c.getTime() + expiry);

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 4]
```diff
@@ -211,30 +213,92 @@ private void updateUserAccessTiers(DbUser dbUser, Agent agent) {
...
+    public boolean areCompliant() {
+      return isBypassed() || (isComplete() && !hasExpired());
     }
-    return completionTime != null;
+  }
+
+  private RenewableAccessModuleStatus mkStatus(ModuleNameEnum name, ModuleTimes times) {
+    return new RenewableAccessModuleStatus()
+        .moduleName(name)
+        .expirationEpochMillis(times.getExpiration().map(x -> x.getTime()).orElse(null))
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:264]: can use a method reference here: `Timestamp::getTime`

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 5]
```diff
@@ -211,30 +213,92 @@ private void updateUserAccessTiers(DbUser dbUser, Agent agent) {
...
+            new ModuleTimes(
+                user.getComplianceTrainingCompletionTime(),
+                user.getComplianceTrainingBypassTime())),
+        mkStatus(
+            ModuleNameEnum.DATAUSEAGREEMENT,
+            new ModuleTimes(
+                user.getDataUseAgreementCompletionTime(), user.getDataUseAgreementBypassTime())),
+        mkStatus(
+            ModuleNameEnum.PROFILECONFIRMATION,
+            new ModuleTimes(user.getProfileLastConfirmedTime(), null)),
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:281]: alt: single param constructor?  (I don't know what my preference is)

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 6]
```diff
@@ -211,30 +213,92 @@ private void updateUserAccessTiers(DbUser dbUser, Agent agent) {
...
+        new ModuleTimes(
+            user.getDataUseAgreementCompletionTime(), user.getDataUseAgreementBypassTime());
+    if (!configProvider.get().access.enableDataUseAgreement) {
       return true;
-    } else if (user.getDataUseAgreementSignedVersion() != null
-        && user.getDataUseAgreementSignedVersion() == getCurrentDuccVersion()
-        && isCompleteAndNotExpired(user.getDataUseAgreementCompletionTime())) {
-      // User has signed the most-recent DUCC version.
+    }
+    if (duaTimes.isBypassed()) {
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:294]: This is a lot nicer than the null check, thanks

**Review by jmthibault79 — APPROVED** [review]
Looks good.  Some questions/comments, nothing obligatory.

UserServiceAccessTest would be a good place to add tests for this - it's specifically for these kinds of checks.  Actually, I see that this PR has already affected those tests, so please correct this before merging.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java (hunk 3 above)
Comment (dmohs) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:233]: Very nice!

File: api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java [hunk 7]
```diff
@@ -232,7 +232,7 @@ public void testSimulateUserFlowThroughRenewal() {
         .isEqualTo(TierAccessStatus.ENABLED);
 
     // Simulate time passing, user is no longer compliant
-    PROVIDED_CLOCK.setInstant(START_INSTANT.plus(1, ChronoUnit.DAYS));
+    PROVIDED_CLOCK.setInstant(START_INSTANT.plus(2, ChronoUnit.DAYS));
```
Comment (dmohs) [inline api/src/test/java/org/pmiops/workbench/access/UserServiceAccessTest.java:235]: I believe I modified the logic here slightly in that now the module won't be considered expired until *after* the expiration date, which I think is fine.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java (hunk 1 above)
Comment (dmohs) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:216]: I'll merge this and schedule time for a refactoring discussion.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java [hunk 8]
```diff
@@ -211,30 +213,82 @@ private void updateUserAccessTiers(DbUser dbUser, Agent agent) {
...
     }
-    return completionTime != null;
+
+    public Optional<Timestamp> getExpiration() {
+      if (isBypassed() || !configProvider.get().access.enableAccessRenewal) {
+        return Optional.empty();
+      }
+      Long expiryDays = configProvider.get().accessRenewal.expiryDays;
+      if (expiryDays == null) {
+        throw new RuntimeException("config value accessRenewal.expiryDays.expiryDays is null");
```
Comment (dmohs) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:239]: Also for follow-up: I disagree with this.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java (hunk 5 above)
Comment (dmohs) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:271]: Good thought. Also split.

**Review by dmohs — COMMENTED** [review]
Test comment from curl.

**Review by petesantos — APPROVED** [review]
Looks good.
Every change I can think of seems like premature optimization since the access modules epic will be coming up soon and we will need to think through that a bit

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java (hunk 8 above)
Comment (petesantos) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:239]: Given this is a compliance requirement I agree this should throw if it is not configured correctly. I have had some issues with codacy as well. It is another tool that seems to get in the way a bit.

Comment (jmthibault79) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:239]: I think Codacy's complaint here is about the specific exception - `RuntimeException` is pretty generic.  We generally prefer to use exceptions corresponding to the HTML error codes.  For this one I think 500 makes sense, which I think is `ServerErrorException`

**Review by dmohs — COMMENTED** [review]
Another test comment from curl.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java (hunk 8 above)
Comment (calbach) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:239]: Yep.. agree with the warning in this case. Throwing `RuntimeException` is very rarely the correct thing to do. In this case, `ServerException` would be fine. `Preconditions.checkNotNull(..)` is also a less commonly used pattern in the codebase, but might be clearer. 

@dmohs on Codacy in general, please do not ignore these presubmit errors. This is the static analysis tool that was chosen to meet a compliance requirement. The resolution should either be to address the issue, or start a thread in slack to change the style settings. More of these may pop up in the near term, and I've asked the team to please help reconfigure Codacy to make it work for our codebase. https://pmi-engteam.slack.com/archives/CHRN2R51N/p1620758246168800

Comment (dmohs) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java:239]: @calbach `Preconditions.checkNotNull` matches the intent better, thank you. https://github.com/all-of-us/workbench/pull/5011

Note that I did not ignore the presubmit error, but made a judgment to forgo the requested change and documented this judgment. With respect to compliance, this is almost always accepted behavior.

**Pull request merged**